README.md
# Ruby RabbitMQ Janus | [![Build Status](https://travis-ci.org/dazzl-tv/ruby-rabbitmq-janus.svg?branch=master)](https://travis-ci.org/dazzl-tv/ruby-rabbitmq-janus) [![Gem Version](https://badge.fury.io/rb/ruby_rabbitmq_janus.svg)](https://badge.fury.io/rb/ruby_rabbitmq_janus) [![inline docs](http://inch-ci.org/github/dazzl-tv/ruby-rabbitmq-janus.svg)](http://inch-ci.org/github/dazzl-tv/ruby-rabbitmq-janus) [![Known Vulnerabilities](https://snyk.io/test/github/dazzl-tv/ruby-rabbitmq-janus/badge.svg)](https://snyk.io/test/github/dazzl-tv/ruby-rabbitmq-janus)

Ruby Gem for Janus WebRTC Gateway integration using RabbitMQ message queue

This gem is used to communicate to a server Janus through RabbitMQ software
(Message-oriented middleware). It waiting a messages to Rails API who send to
RabbitMQ server in a queue for janus server. janus processes a message and send
 to RabbitMQ server in a queue for gem. Once the received message is decoded
 and returned through the Rails API.

This gem is product by [Dazzl.tv](http://dazzl.tv)

![Simple diagram](.diagrams.svg)

## Menu

* [How to use](#how-to-use)
  * [Installation](#installation)
  * [Requirements](#requirements)
  * [Configuration](#configuration)
    * [Janus](#janus)
    * [Generators](#generators)
    * [Requests](#requests)
  * [Usage](#usage)
    * [Information](#usage-information)
    * [Standard request](#standard-request)
    * [Admin request](#admin-request)
    * [Listen Janus Event](#listen-janus-event)
* [Upgrade](#upgrade)
* [Development](#development)
  * [RSpec](#rspec-test)
  * [Documentation](#documentation)
    * [Read documentation](#read-documentation)
    * [Generate developer documentation](#generate-developer-documentation)
* [Contributing](#contributing)

## How to use

### Installation

Use rubygem for installing gem in your application. Add in your Gemfile :

```ruby
gem 'ruby_rabbitmq_janus'
```

And then execute :

```linux
bundle
```

Or install it yourself as :

```linux
gem install ruby_rabbitmq_janus
```

Install basic configuration :

```linux
rails g ruby_rabbitmq_janus:install
```

### Requirements

* Janus Gateway : minimum version `0.7.4` [Source Code](https://github.com/meetecho/janus-gateway) - [Documentation](https://janus.conf.meetecho.com/docs/)
* RabbitMQ server

### Configuration

If you want used a customize configuration see [ruby-rabbitmq-janus.yml](config/default.md)

#### Janus

Configure one (or many) Janus server.
[See janus gateway documentation.](https://janus.conf.meetecho.com/docs/)

Configure the queue with rabbit with incremental number. So the first instance
take a name `to-janus-1` and all queues `from` sending in same queue.

#### Generators

Use generator for complete installation :

```ruby
rails g -h
RubyRabbitmqJanus:
  ruby_rabbitmq_janus:configuration     # Generate a custom configuration file.
  ruby_rabbitmq_janus:create_request    # Create an request to json format for RubyRabbitmqJanus transaction.
  ruby_rabbitmq_janus:default_request   # Copy base request file sending to janus in application.
  ruby_rabbitmq_janus:initializer       # Generate a initializer to this gem for rails application.
  ruby_rabbitmq_janus:install           # Install RubyRabbitmqJanus in your Rails application
```

#### Requests

For create an new request is simple. Use a command generator :

```ruby
rails g ruby_rabbitmq_janus:create_request test info 'transaction:<string>,body:{plugins:false}'
      create  config/requests/test/info.json
```

For more explain in requests files see [default requests](config/requests.md).

### Usage

#### Usage information

This gem use rabbitmq for manage request sending to Janus, so see
[documentation Janus](https://janus.conf.meetecho.com/docs/pages.html) and
[documentation Rabbitmq](http://www.rabbitmq.com/documentation.html). This gem
use bunny gem for create connection with rabbitmq. See many guide :
[bunny documentation](http://rubybunny.info/articles/guides.html).

#### Standard Request

```ruby
require 'ruby_rabbitmq_janus'

# Initialize standard object
t = RubyRabbitmqJanus::RRJ.new

# Ask info Janus Instance
t.session_endpoint_public do |transaction|
  transaction.publish_message('base::info')
end

=> @request={"janus"=>"server_info" ... }
```

#### Admin Request

__NOTE: The request authentication `HMAC-Signed token authentication` is not available.__

```ruby
require 'ruby_rabbitmq_janus'

# Initialize admin object
t = RubyRabbitmqJanus::RRJAdmin.new

# Ask info sessions in Janus Instance
options = { instance: 42 }
t.admin_endpoint(options) do |transaction|
  transaction.publish_message('admin::sessions', options)
end

=> @request={"janus"=>"success" ... "sessions"=>[123, 456, 789]}
```

#### Listen Janus Event

```ruby
require 'ruby_rabbitmq_janus'

# Create a class in your Rails application
actions = RubyRabbitmqJanus::ActionEvents.new.action

# Initialize a thread for listen public queue and send class to thread
RubyRabbitmqJanus::Janus::Concurrencies::Event.new.run(@actions)
```

#### Listen Janus Admin Event

```ruby
require 'ruby_rabbitmq_janus'

# Create a class in your Rails application
actions = RubyRabbitmqJanus::ActionAdminEvents.new.action

# Initialize a thread for listen public admin queue and send class to thread
RubyRabbitmqJanus::Janus::Concurrencies::EventAdmin.new.run(@actions)
```

## Upgrade

For upgrade your application read [CHANGELOG.md](CHANGELOG.md)

## Development

### RSpec test

__Use docker for running SPEC__

```linux
# Prepare images
docker-compose build

# Launch tiers service RabbitMQ
docker-compose up -d rabbit

# Launch tiers service Janus (ensure rabbit is started before and READY)
docker-compose up -d janus janus_token

## Excute ##

# Start for MongoID database
docker-compose run gem env MONGO=true bundle exec rake classic
docker-compose run gem env MONGO=true bundle exec rake concurrency

# Start for SQlite3 database
docker-compose run gem env MONGO=false bundle exec rake classic
docker-compose run gem env MONGO=false bundle exec rake concurrency

## OR ##

# Navigate in container an
docker-compose run gem ash
export MONGO=true
bundle exec rake classic
```

NOTE : Use `SPEC_DEBUG=true bundle exec rake classic` for stopping rspec execution when first test fail.

TIPS: for rspec install janus and rabbitmq server configured by default for user
rabbitmq and use plugin echotest for janus server.

Use tags for rspec :

| Describe                                                      | Type            | Name              |
| ------------------------------------------------------------- | --------------- | ---------------   |
| Internaly function                                            | config          |                   |
| Use bunny gem                                                 |                 | rabbit            |
| Test log functions                                            |                 | log               |
| Test configuration function                                   |                 | config            |
| Test Gem contains CONSTANTS                                   |                 | describe          |
| Level request sending to janus (admin monitor API or classic) | level           |                   |
| Request basic.                                                |                 | base              |
| Request with admin right in Janus application.                |                 | admin             |
| Request candidate/jsep                                        |                 | peer              |
| Request JSON sending to Rabbitmq -> Janus                     | request         |                   |
| Test request attach type                                      |                 | attach            |
| Test request type create                                      |                 | create            |
| Test request type detach                                      |                 | detach            |
| Test request type janus info                                  |                 | info              |
| Test request type test                                        |                 | test              |
| Test request type handle list                                 |                 | handles           |
| Test request type sesssion list                               |                 | sessions          |
| Test request type handle information                          |                 | handle_info       |
| Test request type (un)locking debug                           |                 | set_locking_debug |
| Test request type change log level                            |                 | set_log_level     |
| Test request type tokens list                                 |                 | tokens            |
| Test request type destroy session                             |                 | destroy           |
| Test request keepalive                                        |                 | keepalive         |
| Test request type sdp offer                                   |                 | offer             |
| Test request type trickle, send on candidate                  |                 | trickle           |
| Test request type trickles, send array candidate              |                 | trickles          |
| Event(s) thread                                               | event           |                   |
| Test thread public queue                                      |                 | event             |
| Test thread admin queue                                       |                 | event_admin       |

Example usage rspec with tags :
```ruby
rspec --tag --name:config --tag level:base
```

### Documentation

#### Read documentation

The documentation is accessible in [rubydoc](http://www.rubydoc.info/gems/ruby_rabbitmq_janus/)

#### Generate developer documentation

This doc is generated with yard.

```
# Genereate doc
yard
# Launch server
yard server
```

[See Yard Getting Started](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md)

## Contributing

Bug reports and pull requests are welcome on GitHub at
https://github.com/dazzl-tv/ruby-rabbitmq-janus This project is intended to be a
safe, welcoming space for collaboration, and contributors are expected to adhere
to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [4.0.0] - 2021-05-17
### [Fixed]
- Test with ruby 3
- Change compatibility for rails 6 (and activerecord)
- Change compatibility for mongoid 7

## [3.0.1] - 2021-03-09
### [Fixed]
- Raise an exception when transaction have timeout
- Apply error when read config and return an error

### [Deleted]
- Remove useless error classes

## [3.0.0] - 2019-12-12
### [Deleted]
- Remove deprecated methods
- Remove useless error classes

### [Fixed]
- Mark all message :ack in queue public & admin
- Fix all spec
- Admin request set_session_timeout

### [Add]
- Add error class for event/event_admin process
- Errors when no block given in transaction
- Errors when rabbitmq config is not complete
- Add timeout for transaction
- Add timeout connection to RabbitMQ
- Add default level to error message
- Exception for Janus errors
- Read admin queue and interpret response
- SimpleCov (Code coverage for Ruby)

## [2.7.2] - 2019-10-23
### [Deleted]
- Remove useless private method
- Remove keepalive thread
- Simplify model JanusInstance (remove callbacks and methods)
- Remove gem timers
- Remove Publisher/Listener for JanusInstance events. use sidekiq instead

### [Fixed]
- Fix stop thread. Cleanup connection with RabbitMQ.

### [Added]
- Listen queue admin
- Create multi process for listen public queue

## [2.7.1] - 2019-10-15
### [Fixed]
- Fix namespace raise exception rabbit listener

## [2.7.0] - 2019-09-02
### [Changed]
- Update generator RRJ

### [Deleted]
- Remove gem bunny_mock

### [Added]
- Change method start_transaction with session_endpoint_public and session_endpoint_ private
- Change method start_transaction_handle with handle_endpoint_public and handle_endpoint_private
- Create initialize executed with RSpec
- Add option for rspec response template
- Add basic responses for RSpec

### [Fixed]
- Fix initializer with rspec

## [2.6.0] - 2019-07-24
### [Changed]
- Logger has changed (It's configuration in yml configuration)

###[Added]
- Test with Travis for rails 4/5/6
- Add global variable for loading bunny-mock
- Binaries standalone RRJ threads
- Create Publisher/Listener for JanusInstance events
- Add variable environment for listener path in project

## [2.5.1] - 2019-03-14
###[Fixed]
- Replace tiemout instead session_timeout

## [2.5.0] - 2019-03-08
### [Added]
- Add new element replace for admin action (filename, folder, truncate ...)
- Add missing request admin [see requests](https://janus.conf.meetecho.com/docs/admin.html)
- Add rubocop-rspec (for development)
- Add timeout for spec request
- Use cache for bundler and rvm

### [Changed]
- Rename file request admin handles/session to list_handles/sessions
- Use exclusive queue for admin request
- Update rubocop (for development)
- Change base image for travis execution

### [Fixed]
- Fix version used for sqlite3

## [2.4.0] - 2019-01-31
### [Added]
- Add listener admin queue
- Add Task Admin for CI/sidekiq/Console

## [2.3.0] - 2018-01-25
### [Changed]
- Disabling an instance model now just detach the thread, that will die by itself

## [2.2.2] - 2017-12-21
### [Changed]
- Change configuration for travis, update image trusty
- Add test for ruby 2.4.{1,2,3}

### [Fixed]
- Update gem (nokogiri and yard)

## [2.2.1] - 2017-09-15
### [Fixed]
- Fix restart session and read message in keepalive
- Close connection with rabbitmq when thread it's killed
- Fix name to exception in publisher

## [2.2.0] - 2017-09-14
### [Changed]
- Janus Instance use id field for identify instance
- Optimize dockerfile

### [Added]
- Change option for default cluster mode enabled
- Recreate session when error in thread keepalive
- Add method for test if Janus response it's a error
- Test Janus instance is down and recreate when instance is start
- Restart automatically session when start
- Restart session when ack it's an error
- Timeout for session created
- Add validation to model JanusInstance
- Don't start session initialisation when start rails console
- Add aliases for Janus Instance :
  - session -> session_id
  - thread -> thread_id
  - id -> instance
- spec for Janus Instance model

### [Fixed]
- Model for ActiveRecord
- Create session only cluster mode is disabled
- Cleaning cluster classes
- Get settings to cluster mode
- Managing thread for session with document
- Kill JanusInstance

## [2.1.1] - 2017-06-14
### [Fixed]
- Fix detach handle to correct instance

## [2.1.0] - 2017-06-13
### [Added]
- Janus Instance management
- Initializer for rake task
- Task for deleting all sessions in database and janus
- Task for deleting one sessions in database and janus
- Add generator migration for active record
- spec for two instances to Janus
- spec for cluster class
- spec for models ActiveRecord
- spec for models Mongoid
- Complete initializer template
- Destroy session when gem is deleted
- RabbitMQ logger level
- Format log for task

### [Changed]
- Use templates for generators initializer

### [Fixed]
- Regex transaction is correct
- Travis execute test for master only (rubocop and reek)
- Generators copy/template files

## [2.0.0] - 2017-03-29
### [Added]
- Admin feature
- Complete documentation with example
- Customize response (standard/admin)
- Filter in log for sensitive data (expection debug mode message)
- Spec for RubyRabbitmqJanus::Tools::Type
- Add request for testing type
- Complete default plugins
- Add type convert_array
- add type "<transaction>"
- Spec for RubyRabbitmqJanus::Messages
- Spec for RubyRabbitmqJanus::Rabbit

### [Changed]
- Session ID management
- Type converting
- Delete many method for simplify gem usage

### [Deleted]
- Delete unless method

## [1.2.9] - 2017-02-14
### [Fixed]
- Keepalive deadlock, fixed using semaphore in place of condition variable

## [1.2.8] - 2017-01-27
### [Fixed]

lib/rrj/init.rb
# frozen_string_literal: true

require 'singleton'
require 'yaml'
require 'json'
require 'securerandom'
require 'bunny'
require 'logger'
require 'key_path'
require 'erb'
require 'rrj/tools/bin/config'
require 'rrj/tools/gem/logger'

Log = RubyRabbitmqJanus::Tools::Logger.create unless defined?(Log)
RubyRabbitmqJanus::Tools::Logger.start
#
# :reek:UtilityFunction
# :reek:MissingSafeMethod { exclude: [ cleanup_connection! ] }

# Primary module for this gem
module RubyRabbitmqJanus
  # @author VAILLANT Jeremy <jeremy.vaillant@dazzl.tv>

  # # RubyRabbitmqJanus - RRJ
  #
  # Initialize RRJ gem and create automatically a session with janus and
  # sending a keepalive message. The Time To Live is configured in yaml
  # configuration file `config/default.yml` or if you a customize config in
  # `config/ruby-rabbitmq-janus.yml`.
  #
  # @!attribute [r] session
  #   @return [Fixnum] Return an session number created when this gem is
  #     instanciate. It's janus who creates the number of the session.
  #
  # @see file:/config/requests.md For more information to type requests used.
  # @see file:/config/default.md For more information to config file used.
  class RRJ
    attr_reader :session

    # Return a new instance of RubyRabbitmqJanus.
    #
    # @example Create a instance to this gem
    #   @rrj = RubyRabbitmqJanus::RRJ.new
    #   => #<RubyRabbitmqJanus::RRJ:0x007 @session=123>
    def initialize
      @option = Tools::Option.new
    end

    # Create a transaction between Apps and Janus in queue public
    #
    # @params [Hash] options
    # @options [String] :instance (mandatory id cluster is enabled)
    # @options [Integer] :session_id
    # @options [Hash] :replace
    # @options [Hash] :add
    #
    # @example Create a session
    #   instance : { 'instance' => 42 }
    #   @rrj.session_endpoint_public(instance) do |transaction|
    #     transaction.publish_message('base::create')
    #   end
    #
    # @example Destroy session in instance
    #   options = { 'instance' => 42, 'session_id' => 71984735765 }
    #   @rrj.session_endpoint_public(options) do |transaction|
    #     transaction.publish_message('base::destroy', options)
    #   end
    #
    # @since 2.7.0
    def session_endpoint_public(options = {})
      session = @option.use_current_session?(options)
      transaction = Janus::Transactions::Session.new(false, session)
      transaction.connect { yield(transaction) }
    end

    # Create a transaction between Apps and Janus in queue private
    #
    # @params [Hash] options
    # @options [String] :instance (mandatory id cluster is enabled)
    # @options [Integer] :session_id
    # @options [Hash] :replace
    # @options [Hash] :add
    #
    # @example Create a session
    #   instance : { 'instance' => 42 }
    #   @rrj.session_endpoint_public(instance) do |transaction|
    #     transaction.publish_message('base::create')
    #   end
    #
    # @example Destroy session in instance
    #   options = { 'instance' => 42, 'session_id' => 71984735765 }
    #   @rrj.session_endpoint_public(options) do |transaction|
    #     transaction.publish_message('base::destroy', options)
    #   end
    #
    # @since 2.7.0
    def session_endpoint_private(options = {})
      session = @option.use_current_session?(options)
      transaction = Janus::Transactions::Session.new(true, session)
      transaction.connect { yield(transaction) }
    end

    # Create a transaction between Apps and Janus in private queue
    #
    # @param [Hash] options
    # @options [String] :instance (mandatory id cluster is enabled)
    # @options [Integer] :session_id (mandatory)
    # @options [Integer] :handle_id
    # @options [Hash] :replace
    # @options [Hash] :add
    #
    # @example Post a offer
    #   options = {
    #     'instance' => 42,
    #     'session_id' => 71984735765,
    #     'handle_id' => 56753748917,
    #     'replace' => {
    #       'sdp' => 'v=0\r\no=[..more sdp stuff..]'
    #     }
    #   }
    #   @rrj.handle_endpoint_public(options) do |transaction|
    #     transaction.publish_message('peer::offer', options)
    #   end
    #
    # @since 2.7.0
    def handle_endpoint_public(options = {})
      session = @option.use_current_session?(options)
      handle = @option.use_current_handle?(options)
      instance = options['instance'] || 1
      transaction = Janus::Transactions::Handle.new(false,
                                                    session,
                                                    handle,
                                                    instance)
      transaction.connect { yield(transaction) }
    end

    # Create a transaction between Apps and Janus
    #
    # @param [Hash] options
    # @options [String] :instance (mandatory id cluster is enabled)
    # @options [Integer] :session_id (mandatory)
    # @options [Integer] :handle_id
    # @options [Hash] :replace
    # @options [Hash] :add
    #
    # @example Post a answer
    #   options = {
    #     'instance' => 42,
    #     'session_id' => 71984735765,
    #     'handle_id' => 56753748917,
    #     'replace' => {
    #       'sdp' => 'v=0\r\no=[..more sdp stuff..]'
    #     }
    #   }
    #   @rrj.handle_endpoint_private(options) do |transaction|
    #     transaction.publish_message('peer::answer', options)
    #   end
    #
    # @since 2.7.0
    def handle_endpoint_private(options = {})
      session = @option.use_current_session?(options)
      handle = @option.use_current_handle?(options)
      instance = options['instance'] || 1
      transaction = Janus::Transactions::Handle.new(true,
                                                    session,
                                                    handle,
                                                    instance)
      transaction.connect { yield(transaction) }
    end

    # Delete all resources to JanusInstance reference.
    # Warning: All data in database and Janus Instance is delete
    #
    # @since 2.1.0
    def cleanup_connection!
      Models::JanusInstance.destroys
    end

    private

    attr_reader :option
  end
end
